ntoskrnl/ke/amd64/stubs.c

```diff
@@ -86,26 +91,156 @@ KeZeroPages(IN PVOID Address,
     RtlZeroMemory(Address, Size);
 }
 
+PVOID
+KiSwitchKernelStackHelper(
+    LONG_PTR StackOffset,
+    PVOID OldStackBase);
+
 PVOID
 NTAPI
 KeSwitchKernelStack(PVOID StackBase, PVOID StackLimit)
 {
-    UNIMPLEMENTED;
-    __debugbreak();
-    return NULL;
+    PKTHREAD CurrentThread;
+    PVOID OldStackBase;
+    LONG_PTR StackOffset;
+    SIZE_T StackSize;
+
+    /* Get the current thread */
+    CurrentThread = KeGetCurrentThread();
+
+    /* Save the old stack base */
+    OldStackBase = CurrentThread->StackBase;
+
+    /* Get the size of the current stack */
+    StackSize = (ULONG_PTR)CurrentThread->StackBase - CurrentThread->StackLimit;
+    ASSERT(StackSize <= (ULONG_PTR)StackBase - (ULONG_PTR)StackLimit);
+
+    /* Copy the current stack contents to the new stack */
+    RtlCopyMemory((PUCHAR)StackBase - StackSize,
+                  (PVOID)CurrentThread->StackLimit,
+                  StackSize);
+
+    /* Calculate the offset between the old and the new stack */
+    StackOffset = (PUCHAR)StackBase - (PUCHAR)CurrentThread->StackBase;
+
+    /* Disable interrupts while messing with the stack */
+    _disable();
+
+    /* Set the new trap frame */
+    CurrentThread->TrapFrame = (PKTRAP_FRAME)Add2Ptr(CurrentThread->TrapFrame,
+                                                     StackOffset);
+
+    /* Set the new initial stack */
+    CurrentThread->InitialStack = Add2Ptr(CurrentThread->InitialStack,
+                                          StackOffset);
+
+    /* Set the new stack limits */
+    CurrentThread->StackBase = StackBase;
+    CurrentThread->StackLimit = (ULONG_PTR)StackLimit;
+    CurrentThread->LargeStack = TRUE;
+
+    /* Adjust the PCR fields */
+    __addgsqword(FIELD_OFFSET(KPCR, NtTib.StackBase), StackOffset);
+    __addgsqword(FIELD_OFFSET(KIPCR, Prcb.RspBase), StackOffset);
+
+    /* Finally switch RSP to the new stack.
+       We pass OldStackBase to make sure it is not lost. */
+    OldStackBase = KiSwitchKernelStackHelper(StackOffset, OldStackBase);
```

Comment:
Certain compilers with certain options may use rbp as a frame pointer, which your code wouldn't translate.
For that matter, the x86 version translates ebp but not frame pointers (or any variables) that are already saved on stack. How does that work reliably?

Reply:
I changed the code so that it should fix the issue. This is x64 only (because there the logic of returning to the "root" is already there, while on x86 we still rely on "good faith" that the compiler doesn't do totally awkward stuff.